packages/babel-plugin-minify-dead-code-elimination/src/index.js

```diff
@@ -238,10 +247,25 @@ module.exports = ({ types: t, traverse }) => {
             }
 
             if (binding.references === 1 && binding.kind !== "param" && binding.kind !== "module" && binding.constant) {
+
               let replacement = binding.path.node;
               let replacementPath = binding.path;
+              let isAfterReturn = false;
+
               if (t.isVariableDeclarator(replacement)) {
-                replacement = replacement.init;
+
+                forEachPrevSibling(replacementPath, (path) => {
+                  if (t.isReturnStatement(path)) {
```

Comment:
You're right. But then we have a much "deeper" problem:

```js
function foo() {
  (function() {
    alert(function(){ return x }());
  })();
  var x = 1;
}
```

This brings us again to CFG and knowing when an identifier is referenced before declaration (having to traverse entire tree before declaration).

I think in this case it would be simpler to just hoist vars before doing DCE. So this:

```js
function foo() {
  bar = x;
  var x = 1;
}
```

would first become:

```js
function foo() {
  var x;
  bar = x;
  x = 1;
}
```

which could then be transformed to:

```js
function foo() {
  bar = undefined;
}
```